src/v/archival/ntp_archiver_service.cc

```diff
@@ -98,112 +85,220 @@ ntp_archiver::ntp_archiver(
+    }
+
+    while (!_as.abort_requested()) {
+        if (!_parent.is_elected_leader()) {
+            bool shutdown = false;
+            try {
+                vlog(
+                  _rtclog.debug, "sync manifest loop waiting for leadership");
+                co_await _leader_cond.wait();
+            } catch (const ss::broken_condition_variable&) {
+                shutdown = true;
+            }
+
+            if (shutdown || _as.abort_requested()) {
+                // stop() was called
+                vlog(_rtclog.trace, "sync manifest loop shutting down");
+                break;
+            }
+
+            // We were signalled that we became leader: fall through and
+            // start the upload loop.
+        }
+
+        _start_term = _parent.term();
+        vlog(
+          _rtclog.debug, "sync manifest loop starting in term {}", _start_term);
+
+        try {
+            co_await sync_manifest_loop()
+              .handle_exception_type(
+                [](const ss::abort_requested_exception&) {})
+              .handle_exception_type([](const ss::sleep_aborted&) {})
+              .handle_exception_type([](const ss::gate_closed_exception&) {});
+        } catch (const ss::semaphore_timed_out& e) {
+            vlog(
+              _rtclog.warn,
+              "Semaphore timed out in the upload loop: {}. This may be "
+              "due to the system being overloaded. The loop will "
+              "restart.",
+              e);
+        } catch (...) {
+            vlog(
+              _rtclog.error, "upload loop error: {}", std::current_exception());
+        }
+    }
+}
+
+/// Helper for generating topic manifest to upload
+static cloud_storage::manifest_topic_configuration convert_topic_configuration(
+  const cluster::topic_configuration& cfg,
+  cluster::replication_factor replication_factor) {
+    cloud_storage::manifest_topic_configuration result {
+      .tp_ns = cfg.tp_ns,
+      .partition_count = cfg.partition_count,
+      .replication_factor = replication_factor,
+      .properties = {
+        .compression = cfg.properties.compression,
+        .cleanup_policy_bitflags = cfg.properties.cleanup_policy_bitflags,
+        .compaction_strategy = cfg.properties.compaction_strategy,
+        .timestamp_type = cfg.properties.timestamp_type,
+        .segment_size = cfg.properties.segment_size,
+        .retention_bytes = cfg.properties.retention_bytes,
+        .retention_duration = cfg.properties.retention_duration,
+      },
+    };
+    return result;
+}
+
+ss::future<> ntp_archiver::upload_topic_manifest() {
+    auto topic_cfg_opt = _parent.get_topic_config();
+    if (!topic_cfg_opt) {
+        // This is unexpected: by the time partition_manager instantiates
+        // partitions, they should have had their configs loaded by controller
+        // backend.
+        vlog(
+          _rtclog.error,
+          "No topic configuration available for {}",
+          _parent.ntp());
+        co_return;
+    }
```

Comment:
nit: is it considered a bug if we hit this? Maybe swap this for a `vassert()`? Though I suppose it isn't so bad either way since we'll run this in the next iteration of the upload loop.

Reply:
IMO, whatever the reason is it's not a good thing to crash the whole process because of it. The implication of hitting this is not severe. We will not upload the topic manifest this time but it will be reuploaded later.